apps/webapp/src/state/swap.ts

```diff
@@ -74,28 +73,39 @@ export const createSwapSlice = (): SliceCreator<SwapSlice> => (set, get) => {
 };
 
 const assembleRequest = async ({ assetIn, amount, assetOut }: SwapSlice) => {
-  if (assetIn?.accountIndex === undefined || !assetIn.asset || !assetOut?.penumbraAssetId)
-    throw new Error('missing selected tokens');
+  if (assetIn?.value.valueView.case !== 'knownAssetId') throw new Error('unknown denom selected');
+  if (!assetIn.value.valueView.value.metadata?.penumbraAssetId)
+    throw new Error('missing metadata for assetIn');
+  if (assetIn.address.addressView.case !== 'decoded')
+    throw new Error('address in view is not decoded');
+  if (!assetIn.address.addressView.value.index) throw new Error('No index for assetIn address');
+  if (assetOut?.penumbraAssetId === undefined) throw new Error('assetOut has no asset id');
```

Comment:
This case is sadly tricky because `getDisplayDenomExponent()` below requires the _actual_ `Metadata` type. After zod narrowing, it becomes something different. Let's save this for Thursday's discussion.

Reply:
Another point on `validateAndReturnOrThrow`. This is the equivalent of `Schema.parse(obj)`. And the reason why we aren't using this everywhere is because `validateSchema()` was supposed to be a bit more graceful for our users (only throws in dev, just logs in prod). For invalid schemas, should we throw for everyone though as the default? If so, we can use the standard zod parse.

Reply:
Ahh, gotcha — didn't know that re: `Schema.parse`